core/src/main/java/io/questdb/cairo/O3PurgeDiscoveryJob.java

```diff
@@ -28,181 +28,266 @@
             sink[i] = new StringSink();
             fileNameSinks[i] = new StringSink();
-            txnList[i] = new LongList();
+            partitionList.add(new DirectLongList(configuration.getPartitionPurgeListCapacity() * 2, MemoryTag.NATIVE_LONG_LIST));
+            txnScoreboards.add(new TxnScoreboard(configuration.getFilesFacade(), configuration.getTxnScoreboardEntryCount()));
+            txnReaders.add(new TxReader(configuration.getFilesFacade()));
         }
     }
 
-    public static boolean discoverPartitions(
+    @Override
+    public void close() throws IOException {
+        Misc.freeObjList(partitionList);
+        Misc.freeObjList(txnReaders);
+        Misc.freeObjList(txnScoreboards);
+    }
+
+    private static void processPartition(
+            FilesFacade ff,
+            Path path,
+            int tableRootLen,
+            TxReader txReader,
+            TxnScoreboard txnScoreboard,
+            long partitionTimestamp,
+            int partitionBy,
+            DirectLongList partitionList,
+            int lo,
+            int hi
+    ) {
+        long lastCommittedPartitionName = txReader.getPartitionNameTxnByPartitionTimestamp(partitionTimestamp);
+        if (lastCommittedPartitionName > -1) {
+            assert hi <= partitionList.size();
+            for (int i = lo + 2, n = hi - 1; i < n; i += 2) {
+                long nextNameVersion = Math.min(lastCommittedPartitionName + 1, partitionList.get(i));
+                long previousNameVersion = partitionList.get(i - 2);
+
+                boolean rangeUnlocked = previousNameVersion < nextNameVersion;
+                for (long txn = previousNameVersion; txn < nextNameVersion; txn++) {
+                    if (txnScoreboard.getActiveReaderCount(txn) > 0) {
+                        rangeUnlocked = false;
+                        break;
+                    }
+                }
+
+                if (rangeUnlocked) {
+                    path.trimTo(tableRootLen);
+                    TableUtils.setPathForPartition(path, partitionBy, partitionTimestamp, false);
+                    // -1 here is to compensate +1 added when partition version parsed from folder name
+                    // See comments of why +1 added there in parsePartitionDateVersion()
+                    TableUtils.txnPartitionConditionally(path, previousNameVersion - 1);
+                    path.slash$();
+
+                    // previousNameVersion can be deleted
+                    LOG.info().
+                            $("purging [path=").$(path)
+                            .$(", nameTxnToRemove=").$(previousNameVersion - 1)
+                            .$(", nameTxnNext=").$(nextNameVersion)
+                            .$(", lastCommittedPartitionName=").$(lastCommittedPartitionName)
+                            .I$();
+                    long errno;
+                    if ((errno = ff.rmdir(path)) == 0) {
+                        LOG.info().
```

Comment:
Minor: it would be better for each line to begin with dot, logging will look less like a staircase

Reply:
fixed in fixed in https://github.com/questdb/questdb/pull/1778/commits/c07fd7aa136747a7679d3865ffbb164535d0ee3d